src/runtime_src/core/tools/common/ReportCu.cpp

```diff
@@ -0,0 +1,192 @@
+/**
+ * Copyright (C) 2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "ReportCu.h"
+#include "core/common/query_requests.h"
+#include "core/common/device.h"
+#include "core/common/utils.h"
+
+namespace qr = xrt_core::query;
+
+enum class cu_stat : unsigned short {
+  usage = 0,
+  addr,
+  stat
+};
+
+/*
+ * xclbin locking
+ */
+struct xclbin_lock
+{
+  xclDeviceHandle m_handle;
+  xuid_t m_uuid;
+
+  xclbin_lock(const xrt_core::device *_dev)
+    : m_handle(_dev->get_device_handle())
+  {
+    auto xclbinid = xrt_core::device_query<xrt_core::query::xclbin_uuid>(_dev);
+
+    uuid_parse(xclbinid.c_str(), m_uuid);
+
+    if (uuid_is_null(m_uuid))
+      throw std::runtime_error("'uuid' invalid, please re-program xclbin.");
+
+    if (xclOpenContext(m_handle, m_uuid, std::numeric_limits<unsigned int>::max(), true))
+      throw std::runtime_error("'Failed to lock down xclbin");
+  }
+
+  ~xclbin_lock(){
+    xclCloseContext(m_handle, m_uuid, std::numeric_limits<unsigned int>::max());
+  }
+};

```

Comment:
I have now seen this locking duplicated many times.  Can it be consolidated?
Alternatively, these 4 lines do the same:

```
  // lock xclbin
  auto uuid = xrt::uuid(xrt_core::device_query<xrt_core::query::xclbin_uuid>(device));
  device->open_context(uuid.get(), -1, true);
  auto at_exit = [] (auto device, auto uuid) { device->close_context(uuid.get(), -1); };
  xrt_core::scope_guard<std::function<void()>> g(std::bind(at_exit, device, uuid));
```

If we consolidate, maybe use these 4 lines rather than the xclbin_lock class, unfortunately the scope guard can't be moved into a shared function because it will release the lock when the function returns, so either return the guard object, or just copy and use the 4 lines in your code.